### DISCUSSION TITLE
Auto-discovery and custom Laravel architecture

### DISCUSSION THREAD

**Discussion by AlexandreGerault** [filamentphp/filament#6827] [Help] [answered]
Hi!

I've setup Laravel Filament on an enterprise project. We're actually using the architecture from Laravel Beyond Crud Spatie's e-book. Thus we have this folders architecture:
```
src
├── App
│   ├── Application.php
│   ├── Console
│   ├── Exceptions
│   ├── Http
│   └── Providers
└── Domain
    ├── Admin
    ├── Auth
    ├── Config
    └── Registration
```
For this we customized the `bootstrap/app.php` file this way:
```php
$app = new Application(
    $_ENV['APP_BASE_PATH'] ?? dirname(__DIR__)
);
$app->useAppPath('src/Application');
```
We also have our `Application` implementation:
```php
class Application extends \Illuminate\Foundation\Application
{
    protected $namespace = 'App\\';
}
```
Now, we want to make Laravel filament to look for stuff in `src/Domain/Admin/{Whatever}` folder. To do this, we have the `config/filament.php` config file:
```php
<?php

return [
    'path' => env('FILAMENT_PATH', 'admin'),
    'core_path' => env('FILAMENT_CORE_PATH', 'filament'),
    'domain' => env('FILAMENT_DOMAIN'),
    'pages' => [
        'namespace' => 'Domain\\Admin\\Pages',
        'path' => app_path('../Domain/Admin/Pages'),
        'register' => [
            Pages\Dashboard::class,
        ],
    ],
    'resources' => [
        'namespace' => 'Domain\\Admin\\Resources',
        'path' => app_path('../Domain/Admin/Resources'),
        'register' => [
        ],
    ],
    'widgets' => [
        'namespace' => 'Domain\\Admin\\Widgets',
        'path' => app_path('../Domain/Admin/Widgets'),
        'register' => [
            Widgets\AccountWidget::class,
            Widgets\FilamentInfoWidget::class,
        ],
    ],
    'livewire' => [
        'namespace' => 'Domain\\Admin',
        'path' => app_path('../Domain/Admin'),
    ],
];
```
But it doesn't find the files with this. For now we're registering manually the resources but it requires more work than auto-discovery that shouldn't be a problem as we don't spread Admin stuff in multiple places.

The issues seems related to the way `app_path` works and the fact it points to `src/Application` and not `src`. I tried some tricks to the pathes like `'../' . app_path('../Domain/Admin/Resource')` which works for the application, but breaks the tests.

Any idea how to customize these paths? 🙏

**Comment by danharrin** [comment]
`app_path()` is a Laravel helper that you could swap for your own to generate files in `/src` instead? You definitely don't need `../` as the path is generated relative to the app directory, not the current file.

**Reply by AlexandreGerault** [reply]
Yes `app_path` is a Laravel helper indeed. The basic use of `app_path` returns `src/Application`. So if I do `app_path('../Domain')` it returns `src/Application/../Domain`.

**Reply by danharrin** [reply]
OK but I think ../ is breaking this maybe

**Reply by AlexandreGerault** [reply]
Yes it did. I was thinking about implementing an equivalent of `app_path` but it seems that the use of `base_path` is more easy to adopt for now 🤔 We'll stick to it I think
Thanks for your help 🙏 
EDIT: This is how I did: `base_path(src/Domain/Admin/Resources')`

**Comment by pboivin** [comment, marked as answer]
Just want to echo Dan's suggestion. I'm using the same architecture on my current project and ended up adding a `domain_path()` helper:

```php
if (! function_exists('domain_path')) {
    function domain_path($path = '')
    {
        return app()->basePath('src/Domain/' . $path);
    }
}
```

**Reply by AlexandreGerault** [reply]
Looks neat indeed, thanks for the share
